lib/CliftTransforms/LoopDetection.cpp

```diff
@@ -0,0 +1,211 @@
+    revng_assert(Gotos.back()->getBlock() == BodyBlock);
+    revng_assert(std::next(Gotos.back()->getIterator()) == BodyBlock->end());
+    Gotos.back()->erase();
+
+    // The last loop-goto has been erased and is no longer needed:
+    Gotos = Gotos.drop_back(1);
+  } else {
+    // If the last loop-goto is conditional, a break must be inserted at the end
+    // of the loop body.
+
+    mlir::Value BreakLabel = createLabel(Builder, Function, LoopLoc);
+    Loop.setBreakLabel(BreakLabel);
+
+    Builder.setInsertionPointToEnd(BodyBlock);
+    Builder.create<BreakToOp>(LoopLoc, BreakLabel);
+  }
+
+  // If any loop-gotos still exist, they are replaced by continue-to operations
+  // targeting a newly created continue label.
+  if (not Gotos.empty()) {
+    mlir::Value ContinueLabel = createLabel(Builder, Function, LoopLoc);
+    Loop.setContinueLabel(ContinueLabel);
+
+    for (GotoOp Goto : Gotos) {
+      Builder.setInsertionPoint(Goto);
+      Builder.create<ContinueToOp>(Goto.getLoc(), ContinueLabel);
+      Goto->erase();
+    }
+  }
+
+  if (auto BreakAssignment = getNextOp<AssignLabelOp>(Loop)) {
+    for (mlir::Operation *User : BreakAssignment.getLabel().getUsers()) {
+      auto Goto = mlir::dyn_cast<GotoOp>(User);
+      if (not Goto)
+        continue;
+      if (not Loop->isAncestor(Goto))
+        continue;
+      if (not Loop.getBreakLabel())
+        Loop.setBreakLabel(createLabel(Builder, Function, LoopLoc));
+
+      Builder.setInsertionPoint(Goto);
+      Builder.create<BreakToOp>(Goto.getLoc(), Loop.getBreakLabel());
+      Goto->erase();
+    }
+  }
+
+  if (auto ContinueAssignment = getLastOp<AssignLabelOp>(Loop.getBody())) {
+    for (mlir::Operation *User : ContinueAssignment.getLabel().getUsers()) {
+      auto Goto = mlir::dyn_cast<GotoOp>(User);
+      if (not Goto)
+        continue;
+      if (not Loop->isAncestor(Goto))
+        continue;
+      if (not Loop.getContinueLabel())
+        Loop.setContinueLabel(createLabel(Builder, Function, LoopLoc));
+
+      Builder.setInsertionPoint(Goto);
+      Builder.create<ContinueToOp>(Goto.getLoc(), Loop.getContinueLabel());
+      Goto->erase();
+    }
+  }
+}
+
+static void createLoops(FunctionOp Function) {
+  // Maps each label assignment to the backwards gotos targeting it.
+  llvm::MapVector<AssignLabelOp, llvm::SmallVector<GotoOp, 2>> Labels;
+
+  Function->walk([&](mlir::Operation *Op) {
+    if (auto Label = mlir::dyn_cast<AssignLabelOp>(Op)) {
+      Labels.insert({ Label, {} });
+    } else if (auto Goto = mlir::dyn_cast<GotoOp>(Op)) {
+      mlir::Operation *Assignment = Goto.getLabelAssignmentOp();
+      if (auto A = mlir::dyn_cast<AssignLabelOp>(Assignment)) {
+        if (auto Iterator = Labels.find(A); Iterator != Labels.end()) {
+          if (isBackwardGoto(Goto, Iterator->first))
+            Iterator->second.push_back(Goto);
+        }
+      }
+    }
+  });
+
+  for (const auto &[Label, Gotos] : std::views::reverse(Labels)) {
```

Comment:
AFAIU, `Labels` are collected in syntax order, so using `reverse` here guarantees that we're matching loops in a way that processes first those loops whose label appears last in syntax order.

This is pretty important for the inner functioning of the whole thing. So please explain this explicitly in a comment.

Reply:
You've added comments. Nice